capstone_docs/Maelabord.rmd

```diff
@@ -0,0 +1,753 @@
+      # Punktar fyrir gagnrýnenda- og áhorfendaeinkunnir
+      geom_point(aes(y = tomatometer_rating, color = "Gagnrýnendur ", text = critics_label), size = 2, alpha = 0.7) +
+      geom_point(aes(y = audience_rating, color = "Áhorfendur ", text = audience_label), size = 2, alpha = 0.7) +
+      # Trendlínur fyrir gagnrýnenda- og áhorfendaeinkunnir
+      geom_smooth(aes(y = tomatometer_rating, color = "Gagnrýnendur ", linetype = "Trendline"), 
+                  method = "lm", size = 0.5, se = FALSE) +
+      geom_smooth(aes(y = audience_rating, color = "Áhorfendur ", linetype = "Trendline"), 
+                  method = "lm", size = 0.5, se = FALSE) +
+      # Stillir táknskýringu fyrir liti og línutegundir
+      scale_color_manual(values = c("Gagnrýnendur " = "magenta", "Áhorfendur " = "seagreen1")) +
+      scale_linetype_manual(values = c("Trendline" = "dashed")) +
+      labs(title = "Leonardo DiCaprio: Einkunnir gagnrýnenda vs áhorfenda",
+           x = "Útgáfudagur",
+           y = "Einkunn",
+           color = "",
+      linetype = "") +
+      theme_minimal(base_size = 12) +
+      theme(plot.title = element_text(hjust = 0.5, face = "bold"))
+    
+    ggplotly(trend_comparison_plot_dicaprio, tooltip = "text")
+  })
+  
+  # Graf fyrir Kate Winslet
+  output$rating_trend_plot_winslet <- renderPlotly({
+  trend_comparison_plot_winslet <- ggplot(trend_data_winslet(), aes(x = original_release_date)) +
+    # Lóðréttar línur á milli gagnrýnenda- og áhorfendaeinkunna
+    geom_segment(aes(x = original_release_date, xend = original_release_date,
+                     y = tomatometer_rating, yend = audience_rating),
+                 color = "grey", size = 0.5, alpha = 0.5) +
+    # Punktar fyrir gagnrýnenda- og áhorfendaeinkunnir
+    geom_point(aes(y = tomatometer_rating, color = "Gagnrýnendur ", text = critics_label), size = 2, alpha = 0.7) +
+    geom_point(aes(y = audience_rating, color = "Áhorfendur ", text = audience_label), size = 2, alpha = 0.7) +
+    # Trendlínur fyrir gagnrýnenda- og áhorfendaeinkunnir
+    geom_smooth(aes(y = tomatometer_rating, color = "Gagnrýnendur ", linetype = "Trendline"), 
+                method = "lm", size = 0.5, se = FALSE) +
+    geom_smooth(aes(y = audience_rating, color = "Áhorfendur ", linetype = "Trendline"), 
+                method = "lm", size = 0.5, se = FALSE) +
+    # Stillir táknskýringu fyrir liti og línutegundir
+    scale_color_manual(values = c("Gagnrýnendur " = "magenta", "Áhorfendur " = "seagreen1")) +
+    scale_linetype_manual(values = c("Trendline" = "dashed")) +
+    labs(
+      title = "Kate Winslet: Einkunnir gagnrýnenda vs áhorfenda",
+      x = "Útgáfudagur",
+      y = "Einkunn",
+      color = "",
+      linetype = ""  # Breytir linetype í "Skýring"
+    ) +
+    theme_minimal(base_size = 12) +
+    theme(plot.title = element_text(hjust = 0.5, face = "bold"))
+  
+  ggplotly(trend_comparison_plot_winslet, tooltip = "text")
+})

```

Comment:
1. **Endurtekning**:
   - Kóðinn inniheldur mikla endurtekningu þar sem nánast sama ferli er keyrt bæði fyrir Leonardo DiCaprio og Kate Winslet. Þetta er ekki í anda gullnu reglunnar DRY (Don't Repeat Yourself).
   - Lausn: Búðu til almennt fall sem tekur inn nafn leikara sem stiku og skilar samsvarandi gögnunum og grafi. Með því er hægt að sameina kóðann fyrir báða leikara.

2. **Flytja aðgerðir nær SQL-laginu**:
   - Að reikna og vinna með gagnavinnslu, eins og að búa til labels (`critics_label` og `audience_label`), gæti verið flutt nær SQL-laginu. Þetta myndi minnka vinnsluþungann á Shiny-laginu og gera kóðann hreinni.
   - Í stað þess að keyra þessa vinnslu í R fyrir hvern leikarann væri hægt að bæta dálkum eins og `critics_label` og `audience_label` inn í gagnagrunninn sem viðbótartöflur eða útreiknuð View.

3. **Skilgreiningar og hagræðing gagnaflæðis**:
   - Gögn fyrir Leonardo DiCaprio og Kate Winslet eru unnin í sitthvoru `reactive` fallinu. Í stað þess mætti hafa eitt sameinað `reactive` sem vinnur gögn fyrir alla leikara og skilgreinir þau á sama hátt, byggt á stikum.
   - Þetta gæti verið útfært sem eitt `reactive` með leikaranafnið sem inntaksbreytu.

4. **SQL fyrir trendlínur**:
   - Að reikna trendlínur fyrir gagnrýnenda- og áhorfendaeinkunnir er hægt að gera nær gagnagrunninum með t.d. meðaltali og tímabilshópun (e.g., `GROUP BY YEAR(original_release_date)`).
   - Með þessu sparar þú þungavinnslu í R og bætir uppsetningu á SQL-laginu.